src/main/scala/scalaz/stream/Process.scala

```diff
@@ -69,25 +70,33 @@ sealed trait Process[+F[_], +O]
    * users are responsible for ensuring resource safety.
    */
   final def step: HaltOrStep[F, O] = {
-    def go(cur: Process[F,O], stack: Vector[Cause => Trampoline[Process[F,O]]]) : HaltOrStep[F,O] = {
+    @tailrec
+    def go(cur: Process[F,O], stack: Vector[Cause => Trampoline[Process[F,O]]], cnt: Int) : HaltOrStep[F,O] = {
       if (stack.nonEmpty) cur match {
-        case Halt(cause) => go(Try(stack.head(cause).run), stack.tail)
-        case Emit(os) if os.isEmpty => go(Try(stack.head(End).run), stack.tail)
+        case Halt(End) if cnt <= 0 => Step(empty,Cont(stack))
+        case Halt(cause) => go(Try(stack.head(cause).run), stack.tail, cnt -1)
+        case Emit(os) if os.isEmpty => Step(empty,Cont(stack))
         case emt@(Emit(os)) => Step(emt,Cont(stack))
         case awt@Await(_,_) => Step(awt,Cont(stack))
-        case Append(h,st) => go(h, st fast_++ stack)
+        case Append(h,st) => go(h, st fast_++ stack, cnt -1)
       } else cur match {
         case hlt@Halt(cause) => hlt
-        case emt@Emit(os) if (os.isEmpty) => halt0
+        case emt@Emit(os) if os.isEmpty => halt0
         case emt@Emit(os) => Step(emt,Cont(Vector.empty))
         case awt@Await(_,_) => Step(awt,Cont(Vector.empty))
-        case Append(h,st) => go(h,st)
+        case Append(h,st) => go(h,st, cnt -1)
       }
     }
-    go(this,Vector.empty)
+    // at max stay 10 cycles in step, then consult outer process
+    // this gives chance to higher level combinator that may
+    // interrupt during the process execution to interrupt evaluation of
+    // the process every each 10 cycles.
+    go(this,Vector.empty, 10)
```

Comment:
I think it may be possible to construct an example such that `10` iterations through `go` is not enough. When we jump out of `go` on line 76 (when `cnt <= 0`) then we don't run `stack.head(...)`, which I think means that there may be items on the stack that will never be run, leading to incorrectness. 

The stack does get preserved in `Cont(stack)`, but I'm not convinced this is all we need. Still working on a concrete example of this though.